Handle '>' in attribute values

chklauser opened the pull request:
When looking for the end of a start/empty tag, we use an alternate
version of `read_until` that employs a small state machine to *not*
match on the end byte ('>') when we are currently in an attribute
value.

Fixes #34

I didn't re-enable the tests mentioned in the issue because they would then fail for other reasons. First because of the comment that seems to be parsed correctly but doesn't appear in the test and then because the CData section doesn't match the expected Text event. `quick-xml`s behaviour seems correct.

chklauser commented:
The updated version includes a mention of the fix in the changelog.

tafia commented on src/lib.rs:
```diff
@@ -873,6 +915,81 @@ fn read_until<R: BufRead>(r: &mut R, byte: u8, buf: &mut Vec<u8>)
+            loop {
+                match bytes.next() {
+                    Some((i, &b)) => {
+                        if state == ElemReadState::Elem {
```
I'd prefer for readability something like
```rust
match (state, b) {
    (ElemReadState::Elem, b) if b == end_byte => { ... },
    (ElemReadState::Elem, b'\'') =>  { ... }, 
    // ...
    (ElemReadState::DoubleQ, b'\"') => { ... }
    _ => (), 
}
```

chklauser commented:
I have rebased the patch and changed the state machine code to
```rust
state = match (state, b) {
    (ElemReadState::Elem, b) if b == end_byte => {
        // only allowed to match `end_byte` while we are in state `Elem`
        buf.extend_from_slice(&available[..i]);
        done = true;
        used = i + 1;
        break;
    },
    (ElemReadState::Elem,  b'\'') => ElemReadState::SingleQ,
    (ElemReadState::Elem, b'\"') => ElemReadState::DoubleQ,

    // the only end_byte that gets us out of state 'SingleQ' is a single quote
    (ElemReadState::SingleQ, b'\'') => ElemReadState::Elem,

    // the only end_byte that gets us out of state 'DoubleQ' is a double quote
    (ElemReadState::DoubleQ, b'\"') => ElemReadState::Elem,

    // all other bytes: no state change
    _ => state,
};
```
That looks much nicer indeed :)

tafia commented:
Thanks!
Publishing v0.5.0
